### ISSUE TITLE
PMD for Salesforce: multi-line statement fragment inclusion in the report is unsatisfactory and inconsistent

### ISSUE THREAD

**Issue by sirephil** [pmd/pmd#5734]
**Is your feature request related to a problem? Please describe.**

Salesforce are slowly switching from Checkmarx to the Code Analyzer, which itself incorporates PMD with various Salesforce-specific rules.

We have found that the fragment of a statement included in the report is the final line in the statement, instead of the first line in the statement, for at least one of these rules: `ApexSOQLInjection`. Other rules sensibly report the first (relevant) line.

For example, when considering this code:

```
		Map<Id, User> allUsersById = new Map<Id, User> ((List<User>) (Database.query('SELECT '
				+ fields.toLocalCSV() + ' FROM User WHERE Id IN :recipients OR '
				+ 'Email IN :recipients OR '
				+ userToContactLookupField + ' IN :recipients')));
```

The SOQL CRUD/FLS rule appropriately includes the first line for the SOQL call (here, the content of the `Database.query`, or with a static query, the line including the `[`). However, the SOQL Injection rule reports the final line of that query (here `+ userToContactLookupField + ' IN :recipients')));`).

Since we are an ISV, we have to submit our packages for periodic security review, and this requires submission of scanner reports and a false positives document.

We have annotated every reported "false positive" for Checkmarx scans by including a trailing comment explaining false positive, on the reported line, so the report self-documents all these false positives (allowing our false positives document to be a simple explanation of our approach, at least for the Apex side of things). However, since the SOQL Injection rule (and perhaps others) inconsistently report different lines, we do not have properly self-documenting reports.

**Describe the solution you'd like**

All Salesforce-specific actions should consistently report the first (relevant) line of a multi-line statement. The specific line reported would, ideally, match those included by Checkmarx.

**Describe alternatives you've considered**

Duplicating our false positive comments so they appear on both the Checkmarx reported and PMD reported lines, but this is a lot of extra work and would likely contribute to our total apex characters limit.

**Additional context**

To be clear, we would very much prefer that all rules consistently report the same line as Checkmarx would, for the same type of issue.

**Comment by sirephil** [comment]
BTW, I wasn't sure if this should be a bug or feature request.

**Comment by sirephil** [comment]
Nudge

**Comment by adangel** [comment]
Thanks for reporting.

In general, I agree with the request - whatever position PMD is reporting, it should be useful, relevant, consistent.
We had similar requests in the past already (see e.g. #5616 #5668 #5669 ).

This issue itself it however a bit too broad/vague to be useful for an actionable task...
Since you already mention one specific rule which reports suboptimal positions, we should concentrate on this: We know for sure, that the rule [ApexSOQLInjection](https://docs.pmd-code.org/latest/pmd_rules_apex_security.html#apexsoqlinjection) has such a problem and this can be fixed. But we can't fix something we don't know (e.g. which other rules have similar problems), that's why this issue is difficult (would never be solved, because it is open ended).

I'd suggest to change this issue to specifically address only the problem for [ApexSOQLInjection](https://docs.pmd-code.org/latest/pmd_rules_apex_security.html#apexsoqlinjection) and create new issues for any other rule that should be improved.

**Comment by sirephil** [comment]
I am happy to focus this one on the one example provided here. I will raise other issues as and when identified.
